Lib/test/test_capi/test_codecs.py

```diff
@@ -780,33 +842,70 @@ def test_codec_lookup_error(self):
         self.assertIs(codec_lookup_error('ignore'), codecs.ignore_errors)
         self.assertIs(codec_lookup_error('replace'), codecs.replace_errors)
         self.assertIs(codec_lookup_error('xmlcharrefreplace'), codecs.xmlcharrefreplace_errors)
+        self.assertIs(codec_lookup_error('backslashreplace'), codecs.backslashreplace_errors)
         self.assertIs(codec_lookup_error('namereplace'), codecs.namereplace_errors)
         self.assertRaises(LookupError, codec_lookup_error, 'unknown')
 
-    def test_codec_error_handlers(self):
-        exceptions = [
-            # A UnicodeError with an empty message currently crashes:
-            # See: https://github.com/python/cpython/issues/123378
-            # UnicodeEncodeError('bad', '', 0, 1, 'reason'),
-            UnicodeEncodeError('bad', 'x', 0, 1, 'reason'),
-            UnicodeEncodeError('bad', 'xyz123', 0, 1, 'reason'),
-            UnicodeEncodeError('bad', 'xyz123', 1, 4, 'reason'),
-        ]
-
-        strict_handler = _testcapi.codec_strict_errors
+    def test_codec_strict_errors_handler(self):
+        handler = _testcapi.codec_strict_errors
+        for exc in self.all_unicode_errors + self.bad_unicode_errors:
+            with self.subTest(handler=handler, exc=exc):
+                self.assertRaises(type(exc), handler, exc)
+
+    def test_codec_ignore_errors_handler(self):
+        handler = _testcapi.codec_ignore_errors
+        all_exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, all_exceptions, bad_exceptions)
+
+    def test_codec_replace_errors_handler(self):
+        handler = _testcapi.codec_replace_errors
+        all_exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, all_exceptions, bad_exceptions)
+
+    def test_codec_xmlcharrefreplace_errors_handler(self):
+        handler = _testcapi.codec_xmlcharrefreplace_errors
+        exceptions = self.unicode_encode_errors
+        bad_exceptions = (
+            self.bad_unicode_errors
+            + tuple(e for e in self.all_unicode_errors if e not in exceptions)
+        )
+        self.do_test_codec_errors_handler(handler, exceptions, bad_exceptions)
+
+    def test_codec_backslashreplace_errors_handler(self):
+        handler = _testcapi.codec_backslashreplace_errors
+        exceptions = self.all_unicode_errors
+        bad_exceptions = self.bad_unicode_errors
+        self.do_test_codec_errors_handler(handler, exceptions, bad_exceptions)
+
+    def test_codec_namereplace_errors_handler(self):
+        handler = _testlimitedcapi.codec_namereplace_errors
+        exceptions = self.unicode_encode_errors
+        bad_exceptions = (
+            self.bad_unicode_errors
+            + tuple(e for e in self.all_unicode_errors if e not in exceptions)
+        )
```

Comment:
It seems this expression can be used for `bad_exceptions` in all the test cases.
Could `do_test_codec_errors_handler` compute it?

Reply:
Yes it could. I wasn't sure whether there will be more "specialized" paths to handle in the future, which is why I did a "generic" interface.